### PR TITLE
Promtail: Add new target for receiving GCP PubSub Push messages

### PR DESCRIPTION
**Pull request by thepalbi**
**What this PR does / why we need it**:
Promtail currently supports scraping GCP logs using a [pull model](https://grafana.com/docs/loki/next/clients/promtail/scraping/#gcplog-scraping), in which promtail subscribes to a GCP PubSub topic, to where logs are routed. There's another mechanism possible for receiving messages from PubSub topics (for example logs, but could potentially be other kind of messages) which is by using a [push subscription](https://cloud.google.com/pubsub/docs/push).

This PR adds a new Promtail target which is able to receive messages from a GCP PubSub push subscription. It has been mainly thought for logs, but since it's generic it's able to support whatever message the GCP topic sends, since they all use a similar envelope. The target exposes and HTTP server (with just one `POST /gcp/api/v1/push` endpoint), which is then configured as target of this Push subscriptions.

Just to give an example, the envelope that GCP uses for this messages has the following shape:
```json
{
	"message": {
		"attributes": {
			"logging.googleapis.com/timestamp": "2022-07-25T22:19:09.903683708Z"
		},
		"data": "ewogICJzZXZlcml0eSI6ICJERUJVRyIsCiAgInRleHRQYXlsb2FkIjogIkZ1bmN0aW9uIGV4ZWN1dGlvbiB0b29rIDExOTggbXMuIEZpbmlzaGVkIHdpdGggc3RhdHVzIGNvZGU6IDIwMCIsCiAgInRyYWNlIjogInByb2plY3RzL3dpcmVkLWhlaWdodC90cmFjZXMvNTRhYThhMjA4NzUyNTNjNmI0N2NhYThiZDI4YWQ2NTIiCn0=",
		"messageId": "5187581549398349",
		"message_id": "5187581549398349",
		"publishTime": "2022-07-25T22:19:15.56Z",
		"publish_time": "2022-07-25T22:19:15.56Z"
	},
	"subscription": "projects/wired-height-350515/subscriptions/test"
}
```
This message has been extracted from a push subscription, listening for logs from a cloud-function.

The target can be configured as follows:
```
  gcp_push:
    server:
      http_listen_address: 0.0.0.0
      http_listen_port: 8080
      log_level: debug
      register_instrumentation: true
    labels: 
      job: gcp_push_logs_test
    use_incoming_timestamp: true
```

This target exposes some of the received "label" as "internal labels" to be re-labeled by the user using the scrape config (much like the syslog target does), and dropped if not used. These are:

- `__gcp_message_id`: message id
- `__gcp_attributes_*`: All attributes read from `.message.attributes` are added as a temporal label, conveniently adapting the label name since it might contain non-loki-supported characters. For example for `logging.googleapis.com/timestamp` it's converted to `__gcp_attributes_logging_googleapis_com_timestamp`.

**Some things to keep in mind**:
- GCP offers some nice delivery garantuees for not overloading the receiving sender like a [sliding window algo with slow start](https://cloud.google.com/pubsub/docs/push#delivery_rate).
- It also offers acks and nacks, and retries, but because of promtail async processing, this are lost when the entry enters the underlying channel towards loki's client. The idea is once WAL-support is added, this will improve vastly.

**Which issue(s) this PR fixes**:
n/a

**Special notes for your reviewer**:
Documentation to be added.

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: clients/pkg/promtail/targets/serverutils/config.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/serverutils/config.go:11]: This functionality was shared across all promtails target spinning up an http server (loki push, heroku and this new gcp push one), so extracted it in a `serverutils` package. Happy to rename it or move it elsewhere.

File: clients/pkg/promtail/targets/gcppush/target_test.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/gcppush/target_test.go:45]: Sample payload obtained by setting up a dummy push subscription receiver that logs everything. The `message.data` was simplified, which is just decoded and forwarded to loki, and it can vary depending on what is forwarded from the subscribed topic.

File: clients/pkg/promtail/targets/gcppush/translation.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/gcppush/translation.go:99]: This functionality mimics what was done in https://github.com/grafana/loki/commit/d4cdc376820867cd88248a9b6874e18b1cdac6e6. The idea is to make this target multitenant-supporting, so one can inject the tenantID from upstream using this reserved header (which is well-kown in loki world) and let the loki-client forward it when doing the remote write.

File: clients/pkg/promtail/targets/gcppush/translation.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/gcppush/translation.go:115]: For reviewer, could use some input in here

File: clients/pkg/promtail/targets/gcppush/translation.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/gcppush/translation.go:26]: For reviewer, could use some input in here

**Comment by kavirajk** [comment]
@thepalbi thanks a ton for adding support for GCP pubsub `push` subscription type. Skimmed through the PR and overall looks great :+1: 

Just wanted to clarify one thing. Can we still keep single target type for GCP(`gcplog`) and just add extra config `subscription_type` to differentiate `push` and `pull`? (default being `pull`)? Rather than adding completely new target type.
I see we use more or less similar config for both except the `server` block for `push` subscription type.

We can keep everything more or less same, except if `subscription_type: push` we create newGCPpushTarget.

wdyt?

File: clients/pkg/promtail/scrapeconfig/scrapeconfig.go
Comment (thepalbi) [inline clients/pkg/promtail/scrapeconfig/scrapeconfig.go:40]: drive by: renamed the main target to use a better case

File: clients/pkg/promtail/targets/gcplog/pull_target.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/gcplog/pull_target.go:70]: nit: maybe this one should also be renamed?

File: clients/pkg/promtail/targets/gcplog/push_translation.go
Comment (thepalbi) [inline clients/pkg/promtail/targets/gcplog/push_translation.go:99]: This functionality mimics what was done in https://github.com/grafana/loki/commit/d4cdc376820867cd88248a9b6874e18b1cdac6e6. The idea is to make this target multitenant-supporting, so one can inject the tenantID from upstream using this reserved header (which is well-kown in loki world) and let the loki-client forward it when doing the remote write.

**Comment by thepalbi** [comment]
> @thepalbi thanks a ton for adding support for GCP pubsub `push` subscription type. Skimmed through the PR and overall looks great 👍
> 
> Just wanted to clarify one thing. Can we still keep single target type for GCP(`gcplog`) and just add extra config `subscription_type` to differentiate `push` and `pull`? (default being `pull`)? Rather than adding completely new target type. I see we use more or less similar config for both except the `server` block for `push` subscription type.
> 
> We can keep everything more or less same, except if `subscription_type: push` we create newGCPpushTarget.
> 
> wdyt?

@kavirajk completely agree. Just merged both targets under the same package. Refactored (mainly renames) a little bit the Pull one so that both of them fit better. Let me know what you think 😀

File: clients/pkg/promtail/targets/gcplog/push_translation.go
Comment (kavirajk) [inline clients/pkg/promtail/targets/gcplog/push_translation.go:116]: Good observation about using loop instead of regexp. But let's may be keep that for later, if that end up as bottleneck.

File: clients/pkg/promtail/targets/gcplog/target.go
Comment (kavirajk) [inline clients/pkg/promtail/targets/gcplog/target.go:36]:
```suggestion
		return nil, fmt.Errorf("invalid subscription type: %s. valid options are 'push' and 'pull'", config.SubscriptionType)
```

File: clients/pkg/promtail/targets/target/target.go
Comment (kavirajk) [inline clients/pkg/promtail/targets/target/target.go:49]: We don't need this anymore right?

File: clients/pkg/promtail/targets/gcplog/pull_target.go
Comment (kavirajk) [inline clients/pkg/promtail/targets/gcplog/pull_target.go:21]: Now `PullTarget` doesn't have to be public correct?.

File: clients/pkg/promtail/targets/gcplog/pull_target.go
Comment (kavirajk) [reply clients/pkg/promtail/targets/gcplog/pull_target.go:70]: this is just returning the struct. why don't we make it as part of `newPullTarget` itself?. Any reason to have it as separate function?

**Review by kavirajk — COMMENTED** [review]
LGTM :tada: thanks @thepalbi for addressing the comments.

Left few minor nits and suggestion. We are good to merge after that.

**Comment by kavirajk** [comment]
@thepalbi also can you also please update the docs accordingly?
https://grafana.com/docs/loki/latest/clients/promtail/gcplog-cloud/

File: clients/pkg/promtail/targets/gcplog/push_translation.go
Comment (thepalbi) [reply clients/pkg/promtail/targets/gcplog/push_translation.go:116]: For sure, will leave the comment in case we need to re-evaluate

File: clients/pkg/promtail/targets/target/target.go
Comment (thepalbi) [reply clients/pkg/promtail/targets/target/target.go:49]: Right!

File: clients/pkg/promtail/targets/gcplog/pull_target.go
Comment (thepalbi) [reply clients/pkg/promtail/targets/gcplog/pull_target.go:21]: Yup, since the actual type returned by the public `NewGCPLogTarget` func is `gcplog.Target`, we can keep the actual target types package private.

**Comment by thepalbi** [comment]
> @thepalbi also can you also please update the docs accordingly? https://grafana.com/docs/loki/latest/clients/promtail/gcplog-cloud/

Yup, wanted to have a final 👍 on the implementation before doing this. Will add the docs so the PR is sounds!

File: clients/pkg/promtail/targets/target/target.go
Comment (thepalbi) [reply clients/pkg/promtail/targets/target/target.go]: Done, reverted all those renames to make everything as backwards compatible as possible.

**Comment by thepalbi** [comment]
@kavirajk added the docs on `configuration` and `scraping` sections as discussed. Also resolved all open comments. Can you take another pass + tag the correct docs person?

**Comment by cstyan** [comment]
Do we have any concerns about this target potentially allowing users to easily hit per stream rate limit issues? It looks to me like it would be very easy to configure many things to all write to one pubsub and have it all pushed to promtail as a single stream. 

Should we consider enforcing some amount of the attributes being used as labels?

**Comment by kavirajk** [comment]
@cstyan that's the perfectly valid concern. 

Even now user can take advantage of relabeling those gcp atrributes.
https://github.com/grafana/loki/blob/pablo%2Fgcp-push-target/clients/pkg/promtail/targets/gcplog/formatter.go#L60-L72

And you are right, user may not be aware about the impact before sending too much messages into single stream.

IMO, good place to add this note is in gcp-cloud setup doc https://grafana.com/docs/loki/latest/clients/promtail/gcplog-cloud/

I think @thepalbi decided to update the cloud setup doc in the follow up PR. Can you take care of that when creating follow up PR @thepalbi ?

**Review by kavirajk — APPROVED** [review]
LGTM :+1: 

@KMiller-Grafana can you please approve the doc changes?

**Comment by thepalbi** [comment]
@cstyan sorry I missed you comment. That's a valid point you have. In the trails I ran with actual GCP subscription, I couldn't find any common attribute that might be interested to add to each stream. I know parsing the actual log line (which is always json) can lead to several useful ones. More than happy to:
1. Document this as @kavirajk mentioned
2. If we find some common pattern like a pipeline opening the json and grabbing some field, or unmarshalling directly in the target address it in a follow up

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:334]: Maybe it's obvious, but I'd like to spell out what GCP means.
```suggestion
# Configuration describing how to pull/receive Google Cloud Platform (GCP) logs.
[gcplog: <gcplog_config>]
```

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:943]: Thanks for accepting all these small changes! I'm also copying spelling of "Pub/Sub" from Google Cloud documentation.
```suggestion
The `gcplog` block configures how Promtail receives GCP logs. There are two strategies, based on the configuration of `subscription_type`:
- **Pull**: Using GCP Pub/Sub [pull subscriptions](https://cloud.google.com/pubsub/docs/pull). Promtail will consume log messages directly from the configured GCP Pub/Sub topic.
- **Push**: Using GCP Pub/Sub [push subscriptions](https://cloud.google.com/pubsub/docs/push). Promtail will expose an HTTP server, and GCP will deliver logs to that server.

When using the `push` subscription type, keep in mind:
```

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:945]:
```suggestion
- The `server` configuration is the same as [server](#server), since Promtail exposes an HTTP server for the target that requires it to be the same.
- An endpoint at `POST /gcp/api/v1/push`, which expects requests from a GCP Pub/Sub message delivery system.
```

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:958]: Since we're using the square brackets ([]) to indicate optional configuration parameters, I'd like to eliminate them from the description prose.  My preference is to state the subscription_type in the sentence, as in this suggestion:
```suggestion
# If the subscription_type is pull,  the GCP project ID
[project_id: <string>]

# If the subscription_type is pull, GCP PubSub subscription from where Promtail will pull logs from
[subscription: <string>]

# If the subscription_type is push, the server configuration options
[server: <server_config>]
```

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:974]: How about if we make this easier to skim for our readers?
```suggestion
**Internal labels available for pull**
```

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:980]: And, same here as a suggestion for push:
```suggestion
**Internal labels available for push**
```

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:983]: This might be easier to parse and read:
```suggestion
- `__gcp_attributes_*`: All attributes read from `.message.attributes` in the incoming push message. Each attribute key is conveniently renamed, since it might contain unsupported characters. For example, `logging.googleapis.com/timestamp` is converted to `__gcp_attributes_logging_googleapis_com_timestamp`.
```

File: docs/sources/clients/promtail/scraping.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/scraping.md:196]: Small suggestions here:
```suggestion
Promtail supports scraping cloud resource logs such as GCS bucket logs, load balancer logs, and Kubernetes cluster logs from GCP.
Configuration is specified in the `gcplog` section, within `scrape_config`.
```

File: docs/sources/clients/promtail/scraping.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/scraping.md:254]: Some small grammar improvements here.
```suggestion
When configuring the GCP Log push target, Promtail will start an HTTP server listening on port `8080`, as configured in the `server`
section. This server exposes the single endpoint `POST /gcp/api/v1/push`, responsible for receiving logs from GCP.

It also supports `relabeling` and `pipeline` stages.
```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]
Thanks for these helpful docs.

File: docs/sources/clients/promtail/configuration.md
Comment (thepalbi) [reply docs/sources/clients/promtail/configuration.md:973]: Sure!

**Comment by thepalbi** [comment]
@KMiller-Grafana addressed all comments, and left a new suggestion in one.

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:971]: I'm approving the docs in this PR even without this change.  Please do make this small change, as it fixes a typo.
```suggestion
When Promtail receives GCP logs, various internal labels are made available for [relabeling](#relabeling). This depends on the subscription type chosen.
```

**Review by KMiller-Grafana — APPROVED** [review]
Thank you for revising the docs.  The docs in this PR look good to me.  I put in a suggestion for a typo fix, but I'm approving now so you won't need a re-review from me.

File: clients/pkg/promtail/targets/gcplog/push_target.go
Comment (jeschkies) [inline clients/pkg/promtail/targets/gcplog/push_target.go:99]: I'm not sure the channel would change often. Why not `h.entires`?

File: clients/pkg/promtail/targets/gcplog/push_target_test.go
Comment (jeschkies) [inline clients/pkg/promtail/targets/gcplog/push_target_test.go:138]: Running the server is quite a handful. What do you think about testing the handler function `push` with a [ResponseRecorder](https://pkg.go.dev/net/http/httptest#ResponseRecorder) instead?

File: clients/pkg/promtail/targets/gcplog/push_translation.go
Comment (jeschkies) [reply clients/pkg/promtail/targets/gcplog/push_translation.go:116]: Sorry. I must disagree. This is a rather low hanging optimization that we will probably not attend afterwards. `util.SnakeCase` is already looping through the string. You could a simple method that also escapes characters.

**Review by jeschkies — CHANGES_REQUESTED** [review]
Nice work!

**Comment by thepalbi** [comment]
Thanks for the look @jeschkies , I'll address the comments in a follow up 😀

File: clients/pkg/promtail/targets/gcplog/push_target.go
Comment (thepalbi) [reply clients/pkg/promtail/targets/gcplog/push_target.go:99]: You mean saving `h.handler.Chan()` as a struct field, and directly accessing it?

Comment (jeschkies) [reply clients/pkg/promtail/targets/gcplog/push_target.go:99]: Sorry. I was out. Yes.

**Pull request merged**
